File: modules/offline-core/Sources/OfflineCore/BlobSource.swift
import Foundation

public final class BlobSource: ByteSource {
    public let fileSize: Int64
    public let contentType: String = "audio/mp4"

    private let trackKey: Data
    private let trackIv: Data
    private let blobPath: String
    private let blobStore: BlobStore

    public init(trackId: String, catalog: OfflineCatalog, blobStore: BlobStore, keyVault: KeyVault) throws {
        guard let row = try catalog.get(trackId: trackId) else { throw OfflineError.trackNotFound }
        self.fileSize = row.sizeBytes
        self.blobPath = row.blobPath
        self.blobStore = blobStore
        self.trackIv = row.trackIv
        self.trackKey = try keyVault.unwrap(ciphertext: row.wrappedKey, nonce: row.wrapNonce)
    }

    public func read(range: Range<Int64>) async throws -> Data {
        let aligned = (range.lowerBound / 16) * 16
        let skip = Int(range.lowerBound - aligned)
        let wireLen = Int(range.upperBound - aligned)
        let blockIndex = Int(aligned / 16)
        let cipher = try blobStore.pread(path: blobPath, offset: aligned, length: wireLen)
        let plain = try AESCTRDecryptor.decrypt(ciphertext: cipher, key: trackKey, baseIv: trackIv, blockIndex: blockIndex)
        let wanted = Int(range.upperBound - range.lowerBound)
        return plain.subdata(in: skip..<(skip + wanted))
    }

    public func close() {
        // No explicit teardown needed; ARC releases trackKey at deinit.
    }
}

File: modules/offline-core-android/src/main/kotlin/cc/musicme/offline/BlobSource.kt
package cc.musicme.offline

class BlobSource(
    trackId: String,
    catalog: OfflineCatalog,
    private val blobStore: BlobStore,
    keyVault: KeyVault,
) : ByteSource {

    private val row = catalog.get(trackId) ?: throw OfflineError.TrackNotFound
    override val fileSize: Long = row.sizeBytes
    override val contentType: String = "audio/mp4"
    private val trackKey: ByteArray = keyVault.unwrap(row.wrappedKey, row.wrapNonce)

    override suspend fun read(start: Long, endExclusive: Long): ByteArray {
        val aligned = (start / 16) * 16
        val skip = (start - aligned).toInt()
        val wireLen = (endExclusive - aligned).toInt()
        val blockIndex = (aligned / 16).toInt()
        val cipher = blobStore.pread(row.blobPath, aligned, wireLen)
        val plain = AESCTRDecryptor.decrypt(cipher, trackKey, row.trackIv, blockIndex)
        val wanted = (endExclusive - start).toInt()
        return plain.copyOfRange(skip, skip + wanted)
    }

    override fun close() { trackKey.fill(0) }
}
